cdap-watchdog/src/main/java/io/cdap/cdap/logging/appender/LogAppender.java

```diff
@@ -61,13 +65,41 @@ public final void append(ILoggingEvent eventObject) {
       if (loggingContext == null) {
         return;
       }
+      addErrorCodeTags(eventObject);
       addExtraTags(eventObject, loggingContext);
     }
 
     LogMessage logMessage = new LogMessage(eventObject, loggingContext);
     appendEvent(logMessage);
   }
 
+  /**
+   * if event is for an exception deriving from ErrorCodeProvider, capture error group etc.
+   * @param event
+   */
+  private void addErrorCodeTags(ILoggingEvent event) {
+    if (event.getThrowableProxy() != null && event.getThrowableProxy() instanceof ThrowableProxy) {
+      Throwable throwable = ((ThrowableProxy) (event.getThrowableProxy())).getThrowable();
+      // add all error tags from all the inner exceptions
+      // TODO We may also want to do some autodetection centrally, e.g. SocketException indicates it's network
+      while (throwable != null) {
+        if (throwable instanceof ErrorTagProvider) {
+
+          ErrorTagProvider errorCodeThrowable = (ErrorTagProvider) throwable;
+          StringBuilder tagsSb = new StringBuilder();
+          for (ErrorTagProvider.ErrorTag tag: errorCodeThrowable.getErrorTags()) {
+            if (tagsSb.length() > 0) {
+              tagsSb.append(", ");
+            }
+            tagsSb.append(tag.toString());
```

Comment:
don't need to explicitly call toString()

Reply:
Removed.